Implement Gradle-equivalent to @MavenBuild: @GradleBuild

38leinaD opened the pull request:
#### Short description of what this resolves:

Minimalist but working support for @GradleBuild that triggers a Gradle-build
and retrieves the built WAR-file as a Shrinkwrap-archive as deployment.
Currently, it will grab the first WAR-file it finds under build/libs.
This will not work for EAR multi-project but should be sufficient for 95% of use-cases.

As there is no EmbeddedGradle-equivalent to EmbeddedMaven in shrinkwrap/resolver that can run builds; there only is GradleResolverSystem.
I have directly used the Gradle Tooling API in here. In the bigger picture, I assume it would make sense to rely on extended Gradle-support in shrinkwrap/resolver as the @MavenBuild does.

#### Changes proposed in this pull request:

No changes; just additional features.


**Fixes**: #80

lordofthejars commented on arquillian-chameleon-deployments/arquillian-chameleon-gradle-build-deployment/src/test/java/org/arquillian/container/chameleon/deployment/gradle/GradleBuildAutomaticDeploymentTest.java:
```diff
@@ -0,0 +1,23 @@
+        GradleBuildAutomaticDeployment gradle = new GradleBuildAutomaticDeployment();
+        Archive<?> archive = gradle.build(new TestClass(TestCaseClass.class));
+
+        Assert.assertTrue(archive.contains("/WEB-INF/classes/org/arquillian/example/helloworld/GreetingService.class"));
```
we are using assertj in tests, can you update this assertion to be assertj fashion?

38leinaD replied:
done.

lordofthejars commented on arquillian-chameleon-deployments/arquillian-chameleon-gradle-build-deployment/src/main/java/org/arquillian/container/chameleon/deployment/gradle/GradleBuildAutomaticDeployment.java:
```diff
@@ -0,0 +1,73 @@
+		try {
+			final Path artifactDir = projectModel.getBuildDirectory().toPath().resolve("libs");
+			final Path artifactPath = Files.list(artifactDir)
+				.filter(new Predicate<Path>() {
```
It is not necessary to do a new, you can use directly the lambda approach

38leinaD replied:
The parent pom has
<maven.compiler.target>1.7</maven.compiler.target>
    <maven.compiler.source>1.7</maven.compiler.source>
So, i cannot use Java 8 features here.

lordofthejars commented on arquillian-chameleon-deployments/arquillian-chameleon-gradle-build-deployment/src/main/java/org/arquillian/container/chameleon/deployment/gradle/GradleBuildAutomaticDeployment.java:
```diff
@@ -0,0 +1,73 @@
+					}
+				})	
+				.findFirst()
+				.orElseThrow(new Supplier<Exception>() {
```
It is not necessary to do a new, you can use directly the lambda approach

38leinaD replied:
The parent pom has
<maven.compiler.target>1.7</maven.compiler.target>
<maven.compiler.source>1.7</maven.compiler.source>
So, i cannot use Java 8 features here.

lordofthejars commented on arquillian-chameleon-deployments/arquillian-chameleon-gradle-build-deployment/pom.xml:
```diff
@@ -0,0 +1,70 @@
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
```
I think you are not using mockito anywhere so you can remove this dep.

38leinaD replied:
true. removing it.

lordofthejars commented on arquillian-chameleon-deployments/arquillian-chameleon-gradle-build-deployment/src/main/java/org/arquillian/container/chameleon/deployment/gradle/GradleBuild.java:
```diff
@@ -0,0 +1,25 @@
+     * In case of disabling quiet mode, Gradle output is printed to console
+     * @return 
+     */
+    // TODO
```
Remove this comment?

38leinaD replied:
Removed the quiet-parameter altogether for now. Not implemented yet.

38leinaD commented:
I made the changes according to your suggestions. The only one i am not sure about is regarding lambdas even though the repo seems to have source and target set to 1.7.

lordofthejars commented:
Then I am wondering that you should change java.util.function package since
it was introduced in java8

El sáb., 16 jun. 2018 21:08, Daniel Platz <notifications@github.com>
escribió:

> I made the changes according to your suggestions. The only one i am not
> sure about is regarding lambdas even though the repo seems to have source
> and target set to 1.7.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/arquillian/arquillian-container-chameleon/pull/92#issuecomment-397833105>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABcmYevnvwJw_OJBykinJMWSZnsbeHGhks5t9VeegaJpZM4UmxOL>
> .
>

38leinaD commented:
You are right. I god confused by the setup. Source/Traget is 1.7, but the travis task actually uses/requires oracle8. I think it is because the test-frameworks for src/test/java require Java 8, not the actual arquillian-container-chameleon library.
Anyway, i changed it to be Java 7 compatible.

I additionally, also added:
- @GradleBuild(tasks = "build") so you can customize the gradle-task to run
- @GradleBuild(quiet=false) so you can write the gradle-build-output to system.out for debugging.

lordofthejars commented:
Thank you very much for helping us doing Chameleon even better.

38leinaD commented:
Great! Thanks for merging. Is there any plan soon to release a new version/snapshot of the library and publish on maven central?

bartoszmajsak commented:
@38leinaD we will cut the release this week. Thanks again for your contribution.

38leinaD commented:
@bartoszmajsak Any update on the release?
